src/html/fetch-markdown.js

```diff
@@ -10,102 +10,51 @@
  * governing permissions and limitations under the License.
  */
 const client = require('request-promise-native');
-const URI = require('uri-js');
+const { iter } = require('@adobe/helix-shared').sequence;
 const { bail } = require('../helper');
-
-function uri(root, owner, repo, ref, path) {
-  const rootURI = URI.parse(root);
-  const rootPath = rootURI.path;
-  // remove double slashes
-  const fullPath = `${rootPath}/${owner}/${repo}/${ref}/${path}`.replace(
-    /\/+/g,
-    '/',
-  );
-
-  return URI.serialize({
-    scheme: rootURI.scheme,
-    host: rootURI.host,
-    port: rootURI.port,
-    path: fullPath,
-  });
-}
+const { normalizeURL } = require('../utils/util');
 
 /**
- * Fetches the Markdown specified in the action and returns
- * the body of the Markdown document
- * @param {import("../context").Context} ctx some param
- * @param {import("../context").Action} action some other param
+ * Stage of the markdown->html pipeline that actually fetches the content.
+ *
+ * @param {import("../context").Context} ctx
+ * @param {import("../context").Action} action
  */
-async function fetch(
-  { content: { sources = [] } = {} },
-  {
-    secrets = {},
-    request,
-    logger,
-  },
-) {
+const fetch = async ({ content = [] }, { request, logger }) => {
+  const { sources = [] } = content;
+
   if (!request || !request.params) {
     return bail(logger, 'Request parameters are missing');
   }
 
-  let timeout;
-  if (!secrets.HTTP_TIMEOUT) {
-    logger.warn('No HTTP timeout set, risk of denial-of-service');
-  } else {
-    timeout = secrets.HTTP_TIMEOUT;
-  }
+  const { owner, repo, path, ref } = request.params;
 
-  // get required request parameters
-  const {
-    owner, repo, path,
-  } = request.params;
-
-  let { ref } = request.params;
-
-  // bail if a required parameter cannot be found
-  if (!owner) {
-    return bail(logger, 'Unknown owner, cannot fetch content');
-  }
-  if (!repo) {
-    return bail(logger, 'Unknown repo, cannot fetch content');
-  }
-  if (!path) {
-    return bail(logger, 'Unknown path, cannot fetch content');
-  }
-  if (!ref) {
-    logger.warn(`Recoverable error: no ref given for ${repo}/${owner}.git${path}, falling back to master`);
-    ref = 'master';
+  for (const [key, val] of iter({ owner, repo, path, ref })) {
+    if (!val) {
+      return bail(logger, `Unknown ${key}, cannot fetch content.`);
+    }
   }
 
-  const { REPO_RAW_ROOT: rootPath } = secrets;
-
-  // everything looks good, make the HTTP request
-  const options = {
-    uri: uri(rootPath, owner, repo, ref, path),
-    json: false,
-    timeout,
+  const opt = {
+    url: normalizeURL(`https://api.github.com/repos/${owner}/${repo}/contents/${path}`),
```

Comment:
This breaks local development. GH endpoints should not be hard-coded.  Content might be served from local `git-server`, as e.g. in the following use case:

```bash
hlx demo my-helix-project
cd my-helix-project
hlx up
```

Reply:
Thanks for checking that; I did not test how this behaves with local dev!

This definitely is a problem that needs fixing before this can be merged!
Personally, I don't think just making the domain customizeable  again is a good solution; after all the changes are not just limited to the url; there also are the header changes and there may be more…

So IMO we need a more powerful solution to customize the md fetching step; my first idea would be some `pipe.substitute("fetch", () => custom_impl)` extension point, so we can just provide a custom fetch step in local development…
Together with tooling that makes sure all required repos (as stated in the strain config) are checked out locally, we could get rid of a lot of code and improve the experience for devs working on large projects…

Reply:
You don't have to substitute, just inject your custom fetch before `fetch`, and then the "real fetch" will be skipped because the content is already there.

Reply:
but we pass the `REPO_API_ROOT` which is adjusted to the correct github server during local development. for example:

```
"action": {
  "secrets": {
    "REPO_RAW_ROOT": "http://127.0.0.1:53571/raw/",
    "REPO_API_ROOT": "http://127.0.0.1:53571/api/",
    ...
  },
  "request": {
    "params": {
      "owner": "helix",
      "repo": "github.com--tripodsan--hlxtest",
      "ref": "master",
      "path": "/index.md",
      ...
    },
}
```
So I think if you just use `action.secrets.REPO_API_ROOT`, you're good.